### PR TITLE
Sign & verify with private/public key

### PR DESCRIPTION
**Pull request by RubenMattan**

### CODE REVIEW

File: rogerthat/src/main/java/com/mobicage/rogerthat/EnterPinActivity.java [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+    @Override
+    protected void onServiceBound() {
+    }
+
+    @Override
+    protected void onServiceUnbound() {
+    }
+
+    public void onPinEntered(final String pin) {
+        boolean isValidPin = false;
```
Comment (FastGeert) [inline rogerthat/src/main/java/com/mobicage/rogerthat/EnterPinActivity.java]: Do we have a mechanisme to slow down pin entering when the pin is entered incorrectly several times?

File: rogerthat/src/main/java/com/mobicage/rogerthat/MainActivity.java [hunk 2]
```diff
@@ -144,6 +176,32 @@ private void processIntent() {
...
 
+        if (hasRegistered && AppConstants.SECURE_APP) {
+            try {
+                if (!Security.isPinSet(mService)) {
+                    mBackupIntentAfterPin = intent;
+                    setupPin();
+                    return;
+                }
+            } catch (Exception e) {
+                // todo ruben what should we do here
```
Comment (FastGeert) [inline rogerthat/src/main/java/com/mobicage/rogerthat/MainActivity.java]: todo? this is not finished.

File: rogerthat/src/main/java/com/mobicage/rogerthat/MainActivity.java [hunk 3]
```diff
@@ -144,6 +176,32 @@ private void processIntent() {
...
+                L.bug(e);
+
+                AlertDialog.Builder builder = new AlertDialog.Builder(this);
+                builder.setMessage("Todo ruben: Could not validate security, closing the app.");
```
Comment (FastGeert) [inline rogerthat/src/main/java/com/mobicage/rogerthat/MainActivity.java]: todo? this is not finished.

File: rogerthat/src/main/java/com/mobicage/rogerthat/MainService.java [hunk 4]
```diff
@@ -1498,4 +1517,196 @@ public boolean isPermitted(final String permission) {
         return granted;
     }
 
+    private static class SecurityItem {
+
+        public String uid;
```
Comment (FastGeert) [inline rogerthat/src/main/java/com/mobicage/rogerthat/MainService.java]: Should these instance variables be changeable? If not, make them final.

File: rogerthat/src/main/java/com/mobicage/rogerthat/MainService.java [hunk 5]
```diff
@@ -1498,4 +1517,196 @@ public boolean isPermitted(final String permission) {
...
+
+    public void onPinEntered(final String uid, final String pin) {
+        try {
+            mPrivateKey = Security.getPrivateKey(this, pin);
+            mEnterPinActivityActive = false;
+
+            if (mClearPrivateKeyTimer != null)
+                mClearPrivateKeyTimer.cancel();
+
+            mClearPrivateKeyTimer = new Timer("clear_private_key");
```
Comment (FastGeert) [inline rogerthat/src/main/java/com/mobicage/rogerthat/MainService.java]: Why do we use a timer? Is scheduling a runnable in the future not simpler?

File: rogerthat/src/main/java/com/mobicage/rogerthat/EnterPinActivity.java (hunk 1 above)
Comment (RubenMattan) [inline rogerthat/src/main/java/com/mobicage/rogerthat/EnterPinActivity.java]: No we do not

File: rogerthat/src/main/java/com/mobicage/rogerthat/SetupPinActivity.java [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+
+    @Override
+    protected void onServiceUnbound() {
+    }
+
+    public void onPinEntered(String pin) {
+        if (mFirstPin == null) {
+            mFirstPin = pin;
+            mPinEntryView.clearPinEntry();
+            mMessage.setText("TODO Ruben: Please confirm your pin code.");
```
Comment (FastGeert) [inline rogerthat/src/main/java/com/mobicage/rogerthat/SetupPinActivity.java]: TODO Ruben???

File: rogerthat/src/main/java/com/mobicage/rogerthat/SetupPinActivity.java [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+            return;
+        } else if (mFirstPin.equals(pin)) {
+            try{
+                Security.setPin(mService, pin);
+                setResult(Activity.RESULT_OK);
+                finish();
+                return;
+            } catch (Exception e) {
+                L.d("SetupPinActivity onPinEntered", e);
+                mErrorMessage.setText("TODO Ruben: An unkown error occured while saving the pin code.");
```
Comment (FastGeert) [inline rogerthat/src/main/java/com/mobicage/rogerthat/SetupPinActivity.java]: +1

File: rogerthat/src/main/java/com/mobicage/rogerthat/SetupPinActivity.java [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+            }
+        } else {
+            mErrorMessage.setText("TODO Ruben: Pin code did not match.");
```
Comment (FastGeert) [inline rogerthat/src/main/java/com/mobicage/rogerthat/SetupPinActivity.java]: +1

File: rogerthat/src/main/java/com/mobicage/rogerthat/SetupPinActivity.java [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+        }
+        shakePinEntry();
+        mFirstPin = null;
+        mPinEntryView.clearPinEntry();
+        mMessage.setText("TODO Ruben: to secure your app you need to enter a 4 digit pin code.");
```
Comment (FastGeert) [inline rogerthat/src/main/java/com/mobicage/rogerthat/SetupPinActivity.java]: +1

File: rogerthat/src/main/java/com/mobicage/rogerthat/util/Security.java [hunk 10]
```diff
@@ -112,4 +163,131 @@ public static String generateRandomString(int len) {
         return sb.toString();
     }
 
+    private static byte[] getPinAESKey(String pin) {
```
Comment (FastGeert) [inline rogerthat/src/main/java/com/mobicage/rogerthat/util/Security.java]: Methodname is chosen very badly. Should be getPinMD5Hash()

**Pull request merged**